src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncSpillTests.cs

```diff
@@ -3888,5 +3887,2228 @@ class AltBoolean
+    IL_0091:  stfld      ""int Program.<Assign>d__0.<>1__state""
+    IL_0096:  ldarg.0
+    IL_0097:  ldflda     ""System.Runtime.CompilerServices.AsyncTaskMethodBuilder Program.<Assign>d__0.<>t__builder""
+    IL_009c:  ldloc.3
+    IL_009d:  call       ""void System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetException(System.Exception)""
+    IL_00a2:  leave.s    IL_00b7
+  }
+  IL_00a4:  ldarg.0
+  IL_00a5:  ldc.i4.s   -2
+  IL_00a7:  stfld      ""int Program.<Assign>d__0.<>1__state""
+  IL_00ac:  ldarg.0
+  IL_00ad:  ldflda     ""System.Runtime.CompilerServices.AsyncTaskMethodBuilder Program.<Assign>d__0.<>t__builder""
+  IL_00b2:  call       ""void System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetResult()""
+  IL_00b7:  ret
+}");
+        }
+
+        [Fact]
+        [WorkItem(42755, "https://github.com/dotnet/roslyn/issues/42755")]
+        public void KeepLtrSemantics_ClassFieldAccessOnArray()
+        {
+            var source = @"
+using System;
+using System.Threading.Tasks;
+
+class Program
+{
+    static async Task Assign(A[] arr)
+    {
+        arr[0].x = await Write(""RHS"");
+    }
+
+    static async Task Main(string[] args)
+    {
+        await TestIndexerThrows();
+        await TestAssignmentThrows();
+        await TestIndexerSucceeds();
+        await TestReassignsArrayAndIndexerDuringAwait();
+        await TestReassignsTargetDuringAwait();
+    }
+
+    static async Task TestIndexerThrows()
+    {
+        Console.WriteLine(nameof(TestIndexerThrows));
+        
+        var arr = new A[0];
+        Console.WriteLine(""Before Assignment"");
+        try
+        {
+            await Assign(arr);
+        }
+        catch (IndexOutOfRangeException)
+        {
+            Console.WriteLine(""Caught IndexOutOfRangeException"");
+        }
+    }
+
+    static async Task TestAssignmentThrows()
+    {
+        Console.WriteLine(nameof(TestAssignmentThrows));
+        
+        var arr = new A[1];
+        Console.WriteLine(""Before Assignment"");
+        try
+        {
+            await Assign(arr);
+        }
+        catch (NullReferenceException)
+        {
+            Console.WriteLine(""Caught NullReferenceException"");
+        }
+    }
+
+    static async Task TestIndexerSucceeds()
+    {
+        Console.WriteLine(nameof(TestIndexerSucceeds));
+
+        var arr = new A[1]{ new A() };
+        Console.WriteLine(""Before Assignment arr[0].x is: "" + arr[0].x);
+        await Assign(arr);
+        Console.WriteLine(""After Assignment arr[0].x is: "" + arr[0].x);
+    }
+
+    static async Task TestReassignsArrayAndIndexerDuringAwait()
+    {
+        Console.WriteLine(nameof(TestReassignsArrayAndIndexerDuringAwait));
+
+        var arr = new A[1]{ new A() };
+        var index = 0;
+        Console.WriteLine(""Before Assignment arr.Length is: "" + arr.Length);
+        arr[index].x = await WriteAndReassign(""RHS"");
+        Console.WriteLine(""After Assignment arr.Length is: "" + arr.Length);
```

Comment:
>Console.WriteLine(""After Assignment arr.Length is: "" + arr.Length); [](start = 8, length = 69)

Should assert that the expected element in the original array is modified as we expect.